ath79: glinet_gl-ar750s: Add USB power & microSD

alex2501 opened the pull request:
The GL.iNet AR750S USB and microSD port is currently not working out
of the box.
GPIO 7 is used to control the power of the USB port. Add GPIO 7 as a
fixed-regulator for the port.
Also add &usb1 to DTS to get the microSD port to work.

Signed-off-by: Alexander Wördekemper <alexwoerde@web.de>

jeffsf commented:
Is there a reason to export this rather than use a GPIO hog to pull it up at boot? With a GPIO hog the USB is available in very early boot, allowing use of USB media before user-land processes start to run.

```
+&gpio {
+	usb_power {
+		gpio-hog;
+		gpios = <7 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "usb-power";
+	};
```

Ref: https://patchwork.ozlabs.org/patch/1099811/

alex2501 commented:
No special reason at all. I'm very new to this and just looked how it was done on other devices. Seems yours is the better solution. If we need easier access to control the value, we can still export the GPIO at runtime.

But why do you have GPIO_ACTIVE_HIGH? Thought it must be GPIO_ACTIVE_LOW.

Btw. really nice to see, someone is working on the NAND part.

CHKDSK88 commented:
@alex2501 

Please look this [#1366 (comment)](https://github.com/openwrt/openwrt/pull/1366#issuecomment-424834448).

alex2501 commented:
@CHKDSK88 Thank you. According to this via fixed-regulator should be the best solution.
Tested this and it's working well too. 

@jeffsf Would you like to incorporate the changes into your patches and I will close my pull request? Or should I update my pull request?

ynezz commented:
>  Or should I update my pull request?

Yes, update your pull request as the changes are not related to ar750s NAND support.

Please don't close this PR and open a new one. Please use `git commit --amend` to add changes to your commit. Use `git push --force` to overwrite your github repository/branch. The PR will afterwards update automatically. See the [squashing commits instructions](https://openwrt.org/docs/guide-developer/working-with-github-pr#squashing_commits) on how to modify this PR without closing it and open a new one.

So the result should be just one single commit (you've now two).

jeffsf commented:
I'll be happy to test  a change to using a regulator for `usb-power` both in context of your PR and current source, as well as pending work on the AR750S and future work on the AR300M (which had similar problems on transition to ath79, as I recall).

alex2501 committed:
ath79: glinet_gl-ar750s: Add USB power & microSD

The GL.iNet AR750S USB and microSD port is currently not working out
of the box.
GPIO 7 is used to control the power of the USB port. Add GPIO 7 as a
fixed-regulator for the port.
Also add &usb1 to DTS to get the microSD port to work.

Signed-off-by: Alexander Wördekemper <alexwoerde@web.de>

alex2501 commented:
Thanks for all of your help. Hope this time I got everything correct.

chunkeey commented:
Question: What's the story behind "(must already be inserted at boot)." ?
Is this because your usb<->microsd chip is exporting a usb-storage and it does not indicate
a "medium change" event? (Usually you can still force it by reading/polling /dev/sdX every
once in a while - i.e. after the microsd card gets removed and reinserted . Note: if there's no
poll between a remove and reinsert then I don't think that usb-storage would even notice it.)

jeffsf commented:
To be clear, I'm not a member of the core development team. (ynezz and chunkeey are)

I took a look at this last evening and continue to believe that GPIO 7 is logically ACTIVE_HIGH in terms of providing voltage to the USB port. 

That the OEM firmware defines it as active low is, in my opinion, likely only to use an LED definition as a hack to get the desired boot-time behavior.

https://github.com/gl-inet/openwrt/blob/develop/target/linux/ar71xx/files/arch/mips/ath79/mach-gl-ar750s.c#L91
```
.name           = "gl-ar750s:white:usbpower",
.gpio           = GL_AR750S_GPIO_USB_POWER,
.active_low     = 1,
```

There is no trigger assigned to the LED, so as a result, it is "off" at boot time which the active-low definition inverts to raise the line at boot.

If one examines `/sys/devices/platform/leds-gpio/leds/gl-ar750s:white:usbpower` with this definition, it will be seen after boot that `brightness` is `0` and `trigger` has selected `[none]` and that the USB port is powered. `echo 1 > brightness` will turn _off_ the USB power.

alex2501 commented:
Well, this explanation really makes sense. Updated the PR according to this.

chunkeey commented:
> It appears that media-change events are not presently available and/or are not being recognized and acted on properly.

I know that, but I think this is probably not what you would expect. You see, usb-storage protocol class devices do not have a way to notify the kernel at all through the driver directly. Since if you think about it for usb-sticks and usb-harddrives that spawned the usb-storage class there's no need for something like that, because you can only physically disconnect and reconnect the device, so the physical unplugging and replugging serves as the "media change event".

Now, I was in a similar situation with the WNDR4700, it too has a very similar Genesys USB SD-Card reader GL827L (idVendor=05e3, idProduct=0723 - Product ID is just -1 of yours). So Netgear solved this by connecting the [GL827L](http://www.sunnyqi.com/upLoad/product/month_1308/GL827L.pdf)'s SD_CDZ Pin, which is sort of a interrupt line for sd card detected events to one of the SoC's gpios.
Can you please look if GlInet did the same and check if one of the GPIOs changes when a SD-Card is inserted?

Note: There's an alternative. See: cf8483cb4ffc99bf3f512bb134860ccc8c099abe so check you just need to install the block-mount package on your device and if it works, just add it to the DEVICE_PACKAGES variable.

chunkeey commented:
@alex2501 

can you please word wrap your commit message so that every line stays under 75 characters?
https://openwrt.org/submitting-patches#patch_checklist

alex2501 commented:
@chunkeey Sure, will do this later in the evening.

Just tested with block-mount and it works as aspected. Very nice, will add the package to the DEVICE_PACKAGES in the next commit too.

chunkeey commented:
@alex2501 
I won't be around until next the weekend to merge it. If it can wait that's fine, or somebody else can merge it too.

ynezz commented:
y
